libsignal-service/src/push_service.rs

```diff
@@ -345,6 +388,45 @@ pub trait PushService {
         self.put(&path, messages).await
     }
 
+    /// Request AttachmentV2UploadAttributes
+    ///
+    /// Equivalent with getAttachmentV2UploadAttributes
+    async fn get_attachment_v2_upload_attributes(
+        &mut self,
+    ) -> Result<AttachmentV2UploadAttributes, ServiceError> {
+        self.get("/v2/attachments/form/upload").await
+    }
+
+    /// Upload attachment to CDN
+    ///
+    /// Returns attachment ID and the attachment digest
+    async fn upload_attachment<'s, C: std::io::Read + Send + 's>(
+        &mut self,
+        attrs: &AttachmentV2UploadAttributes,
+        content: &'s mut C,
+    ) -> Result<(u64, Vec<u8>), ServiceError> {
+        let values = [
+            ("acl", &attrs.acl as &str),
+            ("key", &attrs.key),
+            ("policy", &attrs.policy),
+            ("Content-Type", "application/octet-stream"),
+            ("x-amz-algorithm", &attrs.algorithm),
+            ("x-amz-credential", &attrs.credential),
+            ("x-amz-date", &attrs.date),
+            ("x-amz-signature", &attrs.signature),
+        ];
+
+        let mut digester = crate::digeststream::DigestingReader::new(content);
```

Comment:
In the end, if this is just a SHA-256 hasher, why not use it directly (and return the hash from `self.post_to_cdn0`)? I suppose there's a reason that I missed here.

Reply:
Because I'd like all interfaces to be streaming instead of memory-committed. `post_to_cdn0` *probably* cannot be streaming though, so you comment makes sense in that regard.

Otoh, I'd keep the `DigestingReader` because attachment V3 supposedly can be streamed all the way.